torch/distributed/_shard/sharding_plan/api.py

```diff
@@ -0,0 +1,25 @@
+import abc
+import torch.nn as nn
+
+from dataclasses import dataclass
+from typing import Dict
+
+from torch.distributed._shard.sharding_spec import ShardingSpec
+
+@dataclass
+class ShardingPlan(object):
+    """
+    Representation of a sharding plan.
+    Dict keyed by FQN of parameter to PlacementSpec.
+    """
+    plan: Dict[str, ShardingSpec]
```

Comment:
This should be PlacementSpec?

Reply:
I decoupled PlacementSpec with ShardingSpec in the extensible ShardingSpec PR as I don't exactly know how the PlacementSpec base class should behave, maybe a `PlacementSpec` could just be an empty class and ShardingSpec be a subclass of it? (in our proposal a PlacementSpec contains a `apply()-> Tensor` abstract method, which is not applicable to ShardingSpec I think.

Reply:
I think it should be PlacementSpec and we probably need to add a unit test for that too. The reason being that one type of sharding is just placing a Tensor a on gpu 0 and Tensor b on gpu 1 etc.

I think there is a DevicePlacementSpec subclass of PlacementSpec whose `apply` method can essentially just move the Tensor to the appropriate device.